spec/system/admin/orders_spec.rb

```diff
@@ -662,17 +662,77 @@ def extract_pdf_content
             end
           end
 
+          shared_examples "prints invoices accordering to column ordering" do
```

Comment:
I noticed that you often use shared examples for repeated code. That's a totally valid solution but I find them harder to read than defining methods to execute parts of the code.

One reason for this is the implicit passing of variables. The parameter here is `surnames` and you have to read the whole shared example to know that there's this dependency. This would be different when calling a method `expect_invoice_with_surname_order("Adams", "Collins", "Walter")`.

Another reasons is that you can't run one example by line number easily, as in `rspec spec/some_spec.rb:123`.

And lastly, your `before` block contains some of the user actions which, in my opinion, should be part of the `it` block.

I would structure it like this:

```rb
it "order by customer name ascending" do
  page.find('a', text: "NAME").click # orders alphabetically (asc)
  sleep(0.5) # waits for column sorting

  invoice_content = print_all_invoices

  expect(
    invoice_content
  ).to match(/Adams.*Collins.*Walter/m)
end
```

I find the main intention of the user flow easier to read in one code block. The technical details of how to print the invoice are hidden away in a method which could be extracted to helper but can also live in this file for now.

Reply:
Humm, interesting. Thanks so much.

> I find the main intention of the user flow easier to read in one code block. The technical details of how to print the invoice are hidden away in a method which could be extracted to helper but can also live in this file for now.

Code design is a tricky thing, thanks for putting it so clearly. I feel a times it is hard to know what should be sacrificed, DRY vs. code readability. I guess your suggestion combines both :pray: